odoo/addons/base/ir/ir_mail_server.py

```diff
@@ -464,6 +465,70 @@ def send_email(self, message, mail_server_id=None, smtp_server=None, smtp_port=N
             raise MailDeliveryException(_("Mail Delivery Failed"), msg)
         return message_id
 
+    @api.multi
+    def _get_canonical_from(self, original_from):
+        """Returns a FROM header to respect the SMTP server settings.
+
+        Args:
+            original_from (str): The original FROM header for the message.
+
+        Returns:
+            str: A string that should be used in a message header.
+        """
+        # Use the canonical name, if existing.
+        address = parseaddr(original_from)
+        # Otherwise we use the name part of the email address.
+        if not address[0]:
+            address[0] = address[1].split('@', 1)[0]
+        return '%s via %s <%s>' % (
+            address[0],
+            self.env.user.company_id.name,
+            self._get_default_bounce_address(),
+        )
+
+    @api.model
+    def _get_dmarc_whitelist(self):
+        """Return a list of domains that this system is allowed to send from.
+
+        Returns:
+            list(str): Domains that this system can send authenticated emails
+                for.
+        """
+        whitelist = self.env['ir.config_parameter'].sudo().get_param(
+            'mail.dmarc.whitelist',
+        )
+        if whitelist.strip():
+            return [line.strip() for line in whitelist.split(',')]
+
+    @api.multi
+    def _update_message_from(self, message):
+        """Update the FROM header on message to align with server settings.
+
+        Args:
+            message (email.message): The message object to be evaluated and
+                possibly mutated.
+        """
+
+        self.ensure_one()
+
+        domain_catchall = self.env['ir.config_parameter'].sudo().get_param(
+            'mail.catchall.domain',
+        )
+        # Without a catchall domain, we do not have an address to send from
+        if not domain_catchall:
+            return
+
+        whitelist = self._get_dmarc_whitelist()
+        address = parseaddr(message['From'])
```

Comment:
We found `parseaddr()` and `getaddresses()` to be unreliable for malformed cases, as indicated in the comments of [`email_split()`](https://github.com/odoo/odoo/blob/d8e49d3886ab7c5443b9173bd9b3e45e05b72c6c/odoo/tools/mail.py#L473-L494):
```python
>>> parseaddr("Foo")
('', 'Foo')
>>> parseaddr("Foo @Support <foo@bar.com>")
('', 'Foo@Support')
>>> 
```

For stable, it's a good idea to be defensive about that, explicitly or by using `tools.email_split()` whenever you can.


Ideally, we want this patch to have zero effect when the "whitelist" parameter is not defined, as an extra safety against regressions. This could be done easily by first testing the parameter, along with `'mail.catchall.domain'`

Reply:
Good note, thanks! I left the usage in `_get_rewritten_from`, which seems to work fine with the edge cases.